crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,324 @@
+        let keys: Result<_, _> = object_keys.iter().map(|key| bcs::to_bytes(&key)).collect();
+        self.multi_get(OBJECTS_TABLE, keys?).await
+    }
+
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>> {
+        let keys = transactions.iter().map(|tx| tx.inner().to_vec()).collect();
+        self.multi_get(TRANSACTIONS_TABLE, keys).await
+    }
+
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>> {
+        let key = sequence_number.value().to_be_bytes().to_vec();
+        let mut result = self.multi_get(CHECKPOINTS_TABLE, vec![key]).await?;
+        Ok(result.pop())
+    }
+
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()> {
+        let mut items = Vec::with_capacity(objects.len());
+        for object in objects {
+            let object_key = ObjectKey(object.id(), object.version());
+            items.push((bcs::to_bytes(&object_key)?, object));
+        }
+        self.multi_set(OBJECTS_TABLE, items).await
+    }
+
+    async fn save_transactions(&mut self, transactions: &[TransactionData]) -> Result<()> {
+        let items = transactions.iter().map(|transaction| {
+            (
+                transaction.transaction.digest().inner().to_vec(),
+                transaction,
+            )
+        });
+        self.multi_set(TRANSACTIONS_TABLE, items).await
+    }
+
+    async fn save_checkpoint(&mut self, checkpoint: &CheckpointData) -> Result<()> {
+        let summary = &checkpoint.checkpoint_summary;
+        let contents = &checkpoint.checkpoint_contents;
+        let key = summary.sequence_number.to_be_bytes().to_vec();
+        self.multi_set(CHECKPOINTS_TABLE, [(key, (summary, contents))])
+            .await
+    }
+}
+
+impl BigTableClient {
+    pub async fn new(
+        instance_id: String,
+        is_read_only: bool,
+        timeout: Option<Duration>,
+    ) -> Result<Self> {
+        let policy = if is_read_only {
+            "https://www.googleapis.com/auth/bigtable.data.readonly"
+        } else {
+            "https://www.googleapis.com/auth/bigtable.data"
+        };
+        Ok(match std::env::var("BIGTABLE_EMULATOR_HOST") {
+            Ok(emulator_host) => {
+                let auth_channel = AuthChannel {
+                    channel: Channel::from_shared(format!("http://{emulator_host}"))?
+                        .connect_lazy(),
+                    policy: policy.to_string(),
+                    token_provider: None,
+                    token: Arc::new(RwLock::new(None)),
+                };
+                Self {
+                    table_prefix: format!("projects/emulator/instances/{}/tables/", instance_id),
+                    client: BigtableInternalClient::new(auth_channel),
+                }
+            }
+            Err(_) => {
```

Comment:
Consider moving this to a separate initializer that is only used by unit tests/called explicitly when you want to test against the emulator locally. Reduces the risk that someones local run accidentally hits prod

Reply:
agreed, having a separate constructor would be a big win here for readability and usability

Reply:
hitting prod by mistake is really hard. Requires setting up a bunch of stuff locally